src/client/dfuse/pil4dfs/hook.c

```diff
@@ -172,77 +179,88 @@ read_map_file(char **buf, const char str_target[], char **pos)
 static void
 determine_lib_path(void)
 {
-	int   i, rc;
 	char *read_buff_map = NULL;
-	char *pPos = NULL, *pStart = NULL, *pEnd = NULL;
-	char  lib_ver_str[32], lib_dir_str[256];
+	char *pos, *start, *end;
+
+	read_map_file(&read_buff_map);
 
-	read_map_file(&read_buff_map, "/ld-2.", &pPos);
-	if (pPos == NULL) {
-		free(read_buff_map);
-		found_libc = 0;
+	pos = strstr(read_buff_map, "/lib64/ld-linux");
+	if (pos == NULL)
+		/* try a different format */
+		pos = strstr(read_buff_map, "/lib64/ld-2");
+	if (pos == NULL) {
 		printf("Warning: Failed to find ld.so.\n");
-		return;
+		goto err;
 	}
-	for (i = 0; i < 16; i++) {
-		if (strncmp(pPos + i, ".so", 3) == 0) {
-			pEnd = pPos + i + 3;
-			break;
-		}
+	get_string_pos(pos, &start, &end);
+	if ((end - start + 1) >= PATH_MAX) {
+		printf("Warning: path of path_ld is too long.\n");
+		goto err;
 	}
-	if (pEnd == NULL) {
-		printf("Fail to determine the ending position of libc path.\nQuit\n");
-		exit(1);
+	path_ld = malloc(end - start + 1);
+	if (path_ld == NULL) {
+		printf("Warning: Failed to allocate memory for path_ld.\n");
+		goto err;
 	}
-	for (i = 0; i < 100; i++) {
-		if (strncmp(pPos - i, "  /", 3) == 0) {
-			pStart = pPos - i + 2;
-			break;
-		}
+	memcpy(path_ld, start, end - start);
+	path_ld[end - start] = 0;
```

Comment:
Could this not be done with strndup (or D_STRNDUP)?

Reply:
D_STRNDUP is used.